src/irmin-graphql/server.ml

```diff
@@ -227,58 +222,46 @@ module Make_ext(Server: Cohttp_lwt.S.Server)(Config: CONFIG)(Store : Irmin.S)(Pr
-                      | Some s -> Store.Key.v [s]
-                      | None -> Store.Key.empty
-                    in
-                    Lwt.return_ok (Some (tree, key))
-                  )
-              ;
-              io_field "value"
-                ~args:[]
-                ~typ:Presentation.Contents.schema_typ
-                ~resolve:(fun _ (tree, key) ->
-                    Store.Tree.find tree key >|=
-                    Option.map Presentation.Contents.to_src >|=
-                    Result.ok
+                ~args:Arg.[arg "key" ~typ:(non_null Input.key)]
+                ~typ:tree
+                ~resolve:(fun _ (tree, _) key ->
+                    Store.Tree.kind tree key >>= function
+                    | Some `Contents ->
+                      Store.Tree.get_all tree key >|= fun (c, m) ->
+                      Ok (Some (Lazy.(force contents_as_tree (c, m, key))))
+                    | Some `Node ->
+                      Store.Tree.get_tree tree key >|= fun t ->
+                      Ok (Some (Lazy.(force node_as_tree (t, key))))
+                    | None -> Lwt.return_ok None
                   );
-              io_field "metadata"
-                ~args:[]
-                ~typ:Presentation.Metadata.schema_typ
-                ~resolve:(fun _ (tree, key) ->
-                    Store.Tree.find_all tree key >|=
-                    Option.map snd >|=
-                    Option.map Presentation.Metadata.to_src >|=
-                    Result.ok
-                 );
-              io_field "hash"
+              field "hash"
                 ~typ:(non_null string)
                 ~args:[]
-                ~resolve:(fun _ (tree, key) ->
-                    Store.Tree.get_tree tree key >>= fun tree ->
-                    Lwt.return_ok (Irmin.Type.to_string Store.Hash.t (Store.Tree.hash tree))
+                ~resolve:(fun _ (tree, _) ->
+                    let hash = Store.Tree.hash tree in
+                    Irmin.Type.to_string Store.Hash.t hash
                 );
               io_field "tree"
                 ~typ:(non_null (list (non_null tree)))
                 ~args:[]
                 ~resolve:(fun _ (tree, key) ->
-                    Store.Tree.list tree key >>= Lwt_list.map_p (fun (step, kind) ->
+                    Store.Tree.list tree Store.Key.empty >>= Lwt_list.map_p (fun (step, kind) ->
                         let key' = Store.Key.rcons key step in
                         match kind with
-                        | `Contents -> Lwt.return (Lazy.(force contents_as_tree) (tree, key'))
-                        | `Node -> Lwt.return (Lazy.(force node_as_tree) (tree, key'))
+                        | `Contents ->
+                          Store.Tree.get_all tree key' >|= fun (c, m) ->
+                          Lazy.(force contents_as_tree (c, m, key'))
+                        | `Node ->
+                          Store.Tree.get_tree tree key >|= fun t ->
```

Comment:
During testing this line seemed to be raising error when it shouldn't. I tried replacing `key` with `(Store.Key.v [step])` and that gave me the results I was expecting, but maybe that was just coincidental based on my single example.

Reply:
Ah thanks, it's supposed `key'` not `key`.